python/cudf/cudf/tests/test_dataframe.py

```diff
@@ -226,14 +226,16 @@ def test_dataframe_column_add_drop():
     df = DataFrame()
     data = np.asarray(range(10))
     df["a"] = data
-    df["b"] = data
+    df.b = data
```

Comment:
This shouldn't work. This should throw a warning similar to:
```
/home/nfs/kkraus/miniconda3/envs/cudf_dev/bin/ipython:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
```

You should only be able to change data via a `__setattr__` call.

Reply:
I think you need substantially more tests. This logic is -very-high-risk- in that it affects every __attr__ access to a dataframe. I'd like to see tests that:

1. attempt to do what Keith has mentioned here and throw an error ala Pandas
2. valid re-assignments
3. re-re-assignments
4. attempting to `__setattr__` on methods throwing an error
5. attempting to `__setattr__` a valid attribute with invalid data
6. `__setattr__` at the same time as `__getattr__`, can you make them collide?

There are more corner cases here that aren't immediately hopping to mind.

Reply:
Another important test:
7. `__setattr__` on a copy of an object, validating that the `__init__` control works properly.
8. There are ways to copy objects without calling `__init__` on them, we need to discover that case and handle it properly.

Reply:
Agreed that we definitely need more substantial tests. I'm curious about the object copy scenario. It seems unlikely that a dataframe would be copied in such a manner, or that someone would want to add attributes after doing so. It's a duck-typed language, so lots of weird things are possible. Is there precedence for supporting this scenario?

Reply:
Pandas actually issues a warning. I had one going last night, but apparently I deleted it. (I shouldn't have stayed up so late). I'll get the error back in there.

Reply:
@thomcom I've addressed some of your concerns, but I don't see why others need tests (and therefore don't know how to test).

I've added assignment + reassignment loop test, and warning on `__setattr__` for new attributes.

I'm confused how `__setattr__` and `__getattr__` could ever collide, given the GIL, and that this is pure python.
I'm also confused why attempting to __setattr__ on a method would raise that method's error, given that python doesn't invoke it, and our dunder override doesn't either.
Also, since we're forwarding to `__setitem__`, I'm not sure it it make sense to re-test things such as setting a valid attribute with invalid data? (assuming by valid attribute you mean column, otherwise anything is valid)